ts/nni_manager/core/nnimanager.ts

```diff
@@ -476,19 +476,27 @@ class NNIManager implements Manager {
         // TO DO: add CUDA_VISIBLE_DEVICES
         const includeIntermediateResultsEnv = !!(this.config.deprecated && this.config.deprecated.includeIntermediateResults);
 
+        let tunerWs: string;
+        if (globals.args.urlPrefix) {
+            tunerWs = `ws://localhost:${globals.args.port}/${globals.args.urlPrefix}/tuner`;
+        } else {
+            tunerWs = `ws://localhost:${globals.args.port}/tuner`;
+        }
+
         const nniEnv = {
             SDK_PROCESS: 'dispatcher',
             NNI_MODE: mode,
             NNI_CHECKPOINT_DIRECTORY: dataDirectory,
             NNI_LOG_DIRECTORY: getLogDir(),
             NNI_LOG_LEVEL: getLogLevel(),
             NNI_INCLUDE_INTERMEDIATE_RESULTS: includeIntermediateResultsEnv,
+            NNI_TUNER_COMMAND_CHANNEL: tunerWs,
             CUDA_VISIBLE_DEVICES: toCudaVisibleDevices(this.experimentProfile.params.tunerGpuIndices)
         };
         const newEnv = Object.assign({}, process.env, nniEnv);
         const tunerProc: ChildProcess = getTunerProc(command, stdio, newCwd, newEnv);
         this.dispatcherPid = tunerProc.pid!;
-        this.dispatcher = await createDispatcherInterface(tunerProc);
+        this.dispatcher = await createDispatcherInterface();
```

Comment:
so the interfaces of ws and pipe are not the same, right?

Reply:
Yes. There will be no `pipe` anymore.
I plan to use a new command line arg `tuner_command_channel`.
Similar to `dispatcher_pipe`, if `tuner_command_channel` is set, NNI manager will not call `setupTuner()`. Instead, it will only invoke `createDispatcherInterface()` which listens on `ws://localhost:8080/tuner`.
Once NNI manager is initialized, Python experiment code can connect to the URL and starts tuner/strategy logic.